### PR TITLE
Add block directory to inserter menu

### PR DESCRIPTION
**Pull request by talldan**
Moved over from #16524. Props to @ck-lee for all the hard work on the original PR.

## Description

Extend the inserter to show available blocks for download and silently install the block when the block is inserted into the editor

_This PR is still a WIP. Can you help guide me with the implementation and avoiding any pitfalls? I'm new to the block editor._

## Types of changes
- Extend inserter to fetch a list of suggestBlocks similar to reusableBlocks
- The list of suggestBlocks is currently served from a mock API. It is not built yet.
- Extend the filter to show suggestBlocks if there are no results found.
- When "Add" button is clicked, download the block assets from plugin.svn. Then, load and insert the block into the editor without installing it.
- ...

## Relevant Links 
[design] Main thread
https://make.wordpress.org/design/2019/04/26/block-library-installing-blocks-from-within-gutenberg/

[design] Figma
https://www.figma.com/file/QKhoOKXkBN2mHacqvrtdNuI6/Gutenberg-Block-Library%3A-Installation

[design] Mockup feedback
https://make.wordpress.org/design/2019/06/07/block-library-mockups-prototype

[dev] Block Directory Proposal
https://make.wordpress.org/meta/2019/03/08/the-block-directory-and-a-new-type-of-plugin/

[dev] Block Registration
https://github.com/WordPress/gutenberg/blob/master/docs/rfc/block-registration.md

[dev] Block Registration - assets JSON structure
https://github.com/WordPress/gutenberg/pull/13693#issuecomment-491814028

GitHub Block Directory issues
https://github.com/WordPress/block-directory/issues

GitHub Block Directory project
https://github.com/WordPress/block-directory/projects/1

### CODE REVIEW

File: packages/block-editor/src/components/inserter/menu.js [hunk 1]
```diff
@@ -355,9 +357,27 @@ export class InserterMenu extends Component {
...
-						{ isEmpty( suggestedItems ) && isEmpty( reusableItems ) && isEmpty( itemsPerCategory ) && (
-							<p className="editor-inserter__no-results block-editor-inserter__no-results">{ __( 'No blocks found.' ) }</p>
-						) }
+
+						<__experimentalInserterMenuExtension.Slot
+							fillProps={ {
+								onSelect,
+								onHover: this.onHover,
+								filterValue: this.state.filterValue,
+								isMenuEmpty,
```
Comment (gziolo) [inline packages/block-editor/src/components/inserter/menu.js]: Can we rename this prop to better describe it does?

```suggestion
								hasItems,
```

File: packages/block-directory/src/store/controls.js [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+ *
+ * @param {Array}    asset   The url for the JavaScript.
+ * @param {Function} onLoad  Callback function on success.
+ * @param {Function} onError Callback funciton on failure.
+ */
+const loadScript = ( asset, onLoad, onError ) => {
+	if ( ! asset ) {
+		return;
+	}
+	const existing = document.querySelector( `script[src="${ asset.src }"]` );
```
Comment (gziolo) [inline packages/block-directory/src/store/controls.js]: The biggest challenge here is that it doesn't ensure that all dependencies of the script are loaded. It probably won't be the case on the edit post page though unless someone uses one of the WordPress script handles which are not used on this page.

**Comment by gziolo** [comment]
From https://github.com/WordPress/gutenberg/pull/16524#issuecomment-525964556:

> This PR has now been split to 4 PRs in preparation for further review and merging:
>
> Add block directory search settings (#17182)
> Add an extension slot for inserter menu. (#17183)
> Implement REST controller for block directory features. (#17243)
> Implement new block-directory package (#17244)

Should we close all those PRs?

File: packages/block-directory/src/store/controls.js (hunk 2 above)
Comment (noisysocks) [reply packages/block-directory/src/store/controls.js]: Ah, great point. Let's think about this some more and address it in a follow-up PR.

**Review by noisysocks — APPROVED** [review]
Let's get this in behind an experimental flag. Thanks so much to @ck-lee for all your hard work building this, and thanks @talldan for the final push!

**Comment by youknowriad** [comment]
Great work everyone ❤️

**Comment by TimothyBJacobs** [comment]
All these `return WP_Error` need to be `return new WP_Error`, they'll fatal otherwise.

**Comment by ck-lee** [comment]
Thank you, team. @noisysocks, @talldan: You guys are awesome! I appreciate your final push to merge this big one! 

> All these `return WP_Error` need to be `return new WP_Error`, they'll fatal otherwise.
@TimothyBJacobs. I can help fix this as a separate PR.

**Comment by gziolo** [comment]
@ck-lee great work on this PR 🎉 
@noisysocks and @talldan - I appreciate your courage for polishing and merging the big version of PR 💪

File: packages/editor/src/components/inserter-menu-downloadable-blocks-panel/index.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import { debounce } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { __experimentalInserterMenuExtension } from '@wordpress/block-editor';
+import { DownloadableBlocksPanel } from '@wordpress/block-directory';
+import { useState } from '@wordpress/element';
+
+function InserterMenuDownloadableBlocksPanel() {
```
Comment (gziolo) [inline packages/editor/src/components/inserter-menu-downloadable-blocks-panel/index.js]: @noisysocks and @talldan - it looks like all the code from '@wordpress/block-directory' will still get downloaded even when it's disabled for the user with the `GUTENBERG_PHASE` flag. We need to find a way to tree-shake this code for WordPress core.

File: packages/block-directory/src/store/controls.js (hunk 2 above)
Comment (StevenDufresne) [reply packages/block-directory/src/store/controls.js:73]: **Adding an update:**
This problem could potentially be addressed here:
- [Wordpress Core Ticket](https://core.trac.wordpress.org/ticket/48654 ): 
- Gutenberg Ticket: https://github.com/WordPress/gutenberg/issues/21738

File: packages/block-directory/src/components/downloadable-block-header/index.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+			{
+				icon.match( /\.(jpeg|jpg|gif|png)$/ ) !== null ?
+					<img src={ icon } alt="block icon" /> :
+					<span >
+						<BlockIcon icon={ icon } showColors />
+					</span>
+			}
+
+			<div className="block-directory-downloadable-block-header__column">
+				<span role="heading" className="block-directory-downloadable-block-header__title" >
```
Comment (aduth) [inline packages/block-directory/src/components/downloadable-block-header/index.js:25]: The end-to-end AXE tests require an `aria-level` be provided when using `role="heading"`

https://dequeuniversity.com/rules/axe/3.5/aria-required-attr?application=axe-puppeteer

Normally it's not caught in end-to-end tests, but due to some unexpected circumstances, it's become an issue lately (#22712).

What is the intended level here?
Why can't it just be a semantic heading element, vs. `role=`?

See: #22713

**Pull request merged**